### PR TITLE
Changes to Support sending encrypted PROTOBUFF messages

### PR DESCRIPTION
**Pull request by thirumathi**
# Description and Motivation 

This code change enables the simulator to send Protobuff Messages in addition to JSON

# Change type 

- [ ] Bug fix
- [X] New feature
- [X] Enhancement
- [ ] Breaking change (breaks backward compatibility)

**Checklist:**

- [X] All tests passed
- [X] The code follows the code style and conventions of this project
- [X] The change requires a change to the documentation
- [X] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/device-simulation-dotnet/216)

### CODE REVIEW

File: Services/Models/DeviceModel.cs [hunk 1]
```diff
@@ -171,7 +171,10 @@ public enum DeviceModelMessageSchemaFormat
             Text = 10,
 
             [EnumMember(Value = "JSON")]
-            JSON = 20
+            JSON = 20,
+
+            [EnumMember(Value = "PROTO")]
+            PROTO = 30
```
Comment (dluc) [inline Services/Models/DeviceModel.cs]: Please rename this to "Protobuf"
```
[EnumMember(Value = "Protobuf")]
Protobuf = 30
```

File: Services/Models/Proto/GPSPacket.cs [hunk 2]
```diff
@@ -0,0 +1,340 @@
+// <auto-generated>
+//     Generated by the protocol buffer compiler.  DO NOT EDIT!
+//     source: GPSPacket.proto
```
Comment (dluc) [inline Services/Models/Proto/GPSPacket.cs]: Include in the PR also the original .proto file, so users can use it as a starting point for future extensions, and recreate this file too in case of major protobuf changes.

For security reasons, before accepting the PR we'll also have to compare this code with the output generated by protobuf compiler.

File: Services/Services.csproj [hunk 3]
```diff
@@ -6,6 +6,7 @@
     <RootNamespace>Microsoft.Azure.IoTSolutions.DeviceSimulation.Services</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="Google.Protobuf" Version="3.5.1" />
```
Comment (dluc) [inline Services/Services.csproj]: @hathind-ms to confirm licensing

File: Services/data/devicemodels/cab-01-protobuf.json [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{
+  "SchemaVersion": "1.0.0",
+  "Id": "cab-01-proto",
+  "Version": "0.0.1",
+  "Name": "Cab 01 Proto",
+  "Description": "Cab data on Protobuff",
```
Comment (dluc) [inline Services/data/devicemodels/cab-01-protobuf.json]: typo: "protobuff" -> "protobuf". Also would rephrase to "Cab prototype telemetry serialized with protocol buffers"

File: Services/Models/Proto/GPSPacket.cs [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+// </auto-generated>
+#pragma warning disable 1591, 0612, 3021
+#region Designer generated code
+
```
Comment (dluc) [inline Services/Models/Proto/GPSPacket.cs]: in the proto file please add:
```
option csharp_namespace = "Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Protobuf.Models";
```
to avoid risk of collisions when loading classes by name

File: Services/data/devicemodels/cab-01-protobuf.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  "Properties": {
+    "Type": "Cab01Proto",
+    "Location": "Field",
+    "Latitude": 47.445301,
+    "Longitude": -122.296307
+  },
+  "Telemetry": [
+    {
+      "Interval": "00:00:10",
+      "MessageTemplate": "{\"latitude\":${latitude},\"longitude\":${longitude},\"speed\": ${speed},\"heading\": ${heading},\"altitude\":${altitude},\"accuracy\":${accuracy}}",
```
Comment (dluc) [inline Services/data/devicemodels/cab-01-protobuf.json]: this won't be used, and we need a mechanism to point to the right class (see other comments).  We'll revisit the schema, for the time being let's use this field to store the class name (as opposed to hard coding the class name in `DeviceClient.cs`):
```
"MessageTemplate": "Microsoft.Azure.IoTSolutions.DeviceSimulation.Protobuf.Models.GPSPacket",
```

File: Services/data/devicemodels/cab-01-protobuf.json [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      "MessageSchema": {
+        "Name": "GPSPacket;v1",
+        "Format": "PROTO",
```
Comment (dluc) [inline Services/data/devicemodels/cab-01-protobuf.json]: same here, change this to "Protobuf"

File: Services/data/devicemodels/scripts/cab-01-protobuf-state.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    // 0.1 miles around some location
+    var coords = varylocation(center_latitude, center_longitude, 0.1);
+    state.latitude = coords.latitude;
+    state.longitude = coords.longitude;
+
+    // 30 +/- 5%,  Min 0, Max 80
+    state.speed = vary(30, 5, 0, 80);
+
+    // 1 +/- 12%,  Min 0, Max 8
+    state.heading = vary(state.heading, 1, 0, 8);
```
Comment (dluc) [inline Services/data/devicemodels/scripts/cab-01-protobuf-state.js]: either the comment or the code is incorrect, because this will vary by 1% (the second function param)

File: Services/data/devicemodels/scripts/cab-01-protobuf-state.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+    // current altitude +/- 1%,  Min/Max: current +/- 15
+    //state.altitude = vary(state.altitude, 30, state.altitude - 500, state.altitude + 500);
+    state.altitude = vary(state.altitude, 2, state.altitude - 50, state.altitude + 50);
```
Comment (dluc) [inline Services/data/devicemodels/scripts/cab-01-protobuf-state.js]: either the comment or the code is incorrect, because this will vary by 2% (the second function param)

File: Services/data/devicemodels/scripts/cab-01-protobuf-state.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    state.longitude = coords.longitude;
+
+    // 30 +/- 5%,  Min 0, Max 80
+    state.speed = vary(30, 5, 0, 80);
+
+    // 1 +/- 12%,  Min 0, Max 8
+    state.heading = vary(state.heading, 1, 0, 8);
+
+    // current altitude +/- 1%,  Min/Max: current +/- 15
+    //state.altitude = vary(state.altitude, 30, state.altitude - 500, state.altitude + 500);
```
Comment (dluc) [inline Services/data/devicemodels/scripts/cab-01-protobuf-state.js]: pls remove unused code

File: docs/API_SPECS_DEVICE_MODELS.md [hunk 11]
```diff
@@ -1,17 +1,15 @@
 API specifications - Device Models
 ==================================
 
-There are two types of device models:
-1. Stock device models, compiled into the service
-2. Custom device models, created by an application/a user
+## Get list of device models that can be simulated
```
Comment (dluc) [inline docs/API_SPECS_DEVICE_MODELS.md]: could you rebase you branch to the latest master? This would revert many docs changes done recently

File: docs/API_SPECS_SIMULATIONS.md [hunk 12]
```diff
@@ -115,6 +115,10 @@ Content-Type: application/json; charset=utf-8
     {
       "Id": "engine-01",
       "Count": 1
+    },
+    {
+      "Id": "cab-01-proto",
+      "Count": 1
```
Comment (dluc) [inline docs/API_SPECS_SIMULATIONS.md]: the JSON snippets in this files are only examples, so I don't think there's need for any change here

File: Services/DeviceClient.cs [hunk 13]
```diff
@@ -103,13 +105,30 @@ public class DeviceClient : IDeviceClient
 
         public async Task SendMessageAsync(string message, DeviceModel.DeviceModelMessageSchema schema)
         {
-            var eventMessage = new Message(Encoding.UTF8.GetBytes(message));
+            var eventMessage = default(Message);
+
+            switch (schema.Format)
```
Comment (dluc) [inline Services/DeviceClient.cs]: add a default case, throwing an exception about unknown format

File: Services/DeviceClient.cs [hunk 14]
```diff
@@ -103,13 +105,30 @@ public class DeviceClient : IDeviceClient
...
+            {
+                case (DeviceModel.DeviceModelMessageSchemaFormat.JSON):
+                    eventMessage = new Message(Encoding.UTF8.GetBytes(message));
+                    eventMessage.ContentType = "application/json";
+                    eventMessage.ContentEncoding = "utf-8";
+                    break;
+                case (DeviceModel.DeviceModelMessageSchemaFormat.PROTO):
+                    switch (schema.Name)
+                    {
+                        case ("GPSPacket;v1"):
```
Comment (dluc) [inline Services/DeviceClient.cs]: This approach won't scale and would require listing every possible class available.  

As per other comment, use the new `ClassName` device model property to specify which class to use, searching the class in the assembly with reflection.  We should consider the cost of reflection, but that can be done separately to keep this PR small.

1. get class name T from `MessageTemplate`
2. check if T exists, e.g. `typeof(IServicesConfig).GetTypeInfo().Assembly.GetTypes().Where(...)`
3. if T doesn't exists, exception
4. if T exists use JsonConvert.DeserializeObject(message, T)

File: Services/data/devicemodels/cab-01-protobuf.json [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    "Location": "Field",
+    "Latitude": 47.445301,
+    "Longitude": -122.296307
+  },
+  "Telemetry": [
+    {
+      "Interval": "00:00:10",
+      "MessageTemplate": "{\"latitude\":${latitude},\"longitude\":${longitude},\"speed\": ${speed},\"heading\": ${heading},\"altitude\":${altitude},\"accuracy\":${accuracy}}",
+      "MessageSchema": {
+        "Name": "GPSPacket;v1",
```
Comment (dluc) [inline Services/data/devicemodels/cab-01-protobuf.json]: for Protobuf we need a place to store the class name to instantiate the object later in `DeviceClient`. Add a new `ClassName` field here (and relevant classes):
```
"ClassName": "Microsoft.Azure.IoTSolutions.DeviceSimulation.Protobuf.Models.GPSPacket",
```

File: Services/data/devicemodels/cab-01-protobuf.json [hunk 16]
```diff
@@ -0,0 +1,51 @@
+{
```
Comment (dluc) [inline Services/data/devicemodels/cab-01-protobuf.json]: indentation: 4 spaces

File: Services/DeviceClient.cs (hunk 13 above)
Comment (dluc) [reply Services/DeviceClient.cs]: also would refactor out to private methods (nice to have)

File: Services/Utilities.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        public static MethodInfo GetExtensionMethod(string typeName, string assemblyName, string MethodeName)
+        {
+            Tuple<string, string, string> key = new Tuple<string, string, string>(typeName, assemblyName, MethodeName);
+
+            if (methodDictionary.ContainsKey(key))
+            {
+                return methodDictionary[key];
+            }
+
+            List<Type> AssTypes = new List<Type>();
```
Comment (hathind-ms) [inline Services/Utilities.cs]: please rename to assemblyTypes

File: Services/DeviceClient.cs [hunk 18]
```diff
@@ -9,7 +9,13 @@
 using Microsoft.Azure.Devices.Shared;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using Google.Protobuf;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Protobuf.Models;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Runtime;
+using System.Reflection;
```
Comment (avanikp) [inline Services/DeviceClient.cs]: pls sort using statements

File: Services/Exceptions/UnknownMessageFormatException.cs [hunk 19]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions
+{
+    /// <summary>
+    /// This exception is thrown when a client is requesting a resource that
```
Comment (avanikp) [inline Services/Exceptions/UnknownMessageFormatException.cs]: Comment needs to be updated

File: Services/Utilities.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
+﻿using System;
```
Comment (avanikp) [inline Services/Utilities.cs]: missing copyright comment

File: Services/Utilities.cs [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+            AssTypes.AddRange(gpAssembly.GetTypes());
+
+            var query = from type in AssTypes
+                        where type.IsSealed && !type.IsGenericType && !type.IsNested
+                        from method in type.GetMethods(BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic)
+                        where method.IsDefined(typeof(ExtensionAttribute), false)
+                        where method.GetParameters()[0].ParameterType.Name.Equals(typeName, StringComparison.InvariantCultureIgnoreCase)
+                        select method;
+
+            var mi = from methode in query.ToArray<MethodInfo>()
```
Comment (hathind-ms) [inline Services/Utilities.cs]: Did you mean _method_? Also please rename mi to more meaningful variable name.. methodInfo maybe?

File: Services/data/devicemodels/scripts/cab-01-protobuf-state.js [hunk 22]
```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (avanikp) [inline Services/data/devicemodels/scripts/cab-01-protobuf-state.js]: Add unit tests.(Refer to JavascriptInterpreterTest)

File: Services/Utilities.cs [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+{
+    public static class Utilities
+    {
+        static Dictionary<Tuple<string, string, string>, MethodInfo> methodDictionary = new Dictionary<Tuple<string, string, string>, MethodInfo>();
+
+         /// <summary>
+        /// Extends the System.Type-type to search for a given extended MethodeName.
+        /// </summary>
+        /// <param name="MethodeName">Name of the Methode</param>
+        /// <returns>the found Methode or null</returns>
```
Comment (avanikp) [inline Services/Utilities.cs]: pls add documentation comments for other params too

File: Services/Utilities.cs [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+        public static MethodInfo GetExtensionMethod(string typeName, string assemblyName, string MethodeName)
+        {
+            Tuple<string, string, string> key = new Tuple<string, string, string>(typeName, assemblyName, MethodeName);
```
Comment (avanikp) [inline Services/Utilities.cs]: Add validation for input params, early return for null/empty/invalid cases

File: Services/Exceptions/UnknownMessageFormatException.cs (hunk 19 above)
Comment (avanikp) [inline Services/Exceptions/UnknownMessageFormatException.cs]: pls update the comment

File: Services/DeviceClient.cs [hunk 25]
```diff
@@ -103,13 +110,54 @@ public class DeviceClient : IDeviceClient
...
+
+                        MethodInfo methodInfo = Utilities.GetExtensionMethod("imessage", "Google.Protobuf", "ToByteArray");
+                        if (methodInfo != null)
+                        {
+                            object result = methodInfo.Invoke(jsonObj, new object[] { jsonObj });
+                            if (result != null)
+                            {
+                                byte[] byteArray = result as byte[];
+                                eventMessage = new Message(byteArray);
+								eventMessage.Properties.Add(CLASSNAME_PROPERTY, schema.ClassName);
```
Comment (avanikp) [inline Services/DeviceClient.cs]: nit: fix indentation

File: Services/DeviceClient.cs [hunk 26]
```diff
@@ -103,13 +110,54 @@ public class DeviceClient : IDeviceClient
...
+
+                    if (type != null)
+                    {
+                        object jsonObj = JsonConvert.DeserializeObject(message, type);
+
+                        MethodInfo methodInfo = Utilities.GetExtensionMethod("imessage", "Google.Protobuf", "ToByteArray");
+                        if (methodInfo != null)
+                        {
+                            object result = methodInfo.Invoke(jsonObj, new object[] { jsonObj });
+                            if (result != null)
```
Comment (avanikp) [inline Services/DeviceClient.cs]: what if result is null ?

File: Services/DeviceClient.cs [hunk 27]
```diff
@@ -103,13 +110,54 @@ public class DeviceClient : IDeviceClient
...
+                    }
+                    else
+                    {
+                        throw new ResourceNotFoundException($"Type: {schema.ClassName} not found");
+                    }
+
+                    break;
+                default:
+                    throw new UnknownMessageFormatException($"Message format {schema.Format.ToString()} is invalid. Check the Telemetry format against the permitted values Binary, Text, Json, Protobuf");
+					break;
```
Comment (avanikp) [inline Services/DeviceClient.cs]: nit: fix indentation

File: Services/DeviceClient.cs (hunk 13 above)
Comment (avanikp) [reply Services/DeviceClient.cs]: +1 for refactoring into private methods

File: Services/DeviceClient.cs [hunk 28]
```diff
@@ -103,13 +110,54 @@ public class DeviceClient : IDeviceClient
 
         public async Task SendMessageAsync(string message, DeviceModel.DeviceModelMessageSchema schema)
         {
-            var eventMessage = new Message(Encoding.UTF8.GetBytes(message));
+            var eventMessage = default(Message);
+
+            switch (schema.Format)
+            {
+                case (DeviceModel.DeviceModelMessageSchemaFormat.JSON):
```
Comment (avanikp) [inline Services/DeviceClient.cs]: nit: no need for "(" ")"

File: Services/DeviceClient.cs [hunk 29]
```diff
@@ -1,15 +1,18 @@
...
-using System.Text;
-using System.Threading.Tasks;
 using Microsoft.Azure.Devices.Client;
 using Microsoft.Azure.Devices.Shared;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions;
 using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models;
+using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
+using System;
```
Comment (hathind-ms) [inline Services/DeviceClient.cs]: System namespaces should be on top

File: Services/Services.csproj (hunk 3 above)
Comment (hathind-ms) [reply Services/Services.csproj]: I confirmed. We can use it.

File: Services/Simulation/JavascriptInterpreter.cs [hunk 30]
```diff
@@ -57,6 +63,20 @@ public class JavascriptInterpreter : IJavascriptInterpreter
...
+            this.Invoke(filename, null, context, state, properties);
+        }
+
+        /// <summary>
+        /// Load a JS file and execute the main() function, passing in
+        /// context information and the output from the previous execution.
+        /// Modifies the internal device state with the latest values.
+        /// </summary>
+        public void Invoke(
+        string filename, object scriptParams,
```
Comment (dluc) [inline Services/Simulation/JavascriptInterpreter.cs]: indentation

File: Services/data/devicemodels/cab-01-protobuf.json [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+                    "accuracy": "integer"
+                }
+            }
+        },
+        {
+            "Interval": "00:00:30",
+            "MessageTemplate": "{\"x\":${x},\"y\":${y},\"z\": ${z}}",
+            "MessageSchema": {
+                "Name": "Coordinates;v1",
+                "ClassName": "Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Protobuf.Models.Coordinates",
```
Comment (dluc) [inline Services/data/devicemodels/cab-01-protobuf.json]: this class doesn't exist

File: Services/Simulation/JavascriptInterpreter.cs [hunk 32]
```diff
@@ -73,10 +73,10 @@ public class JavascriptInterpreter : IJavascriptInterpreter
...
-        Dictionary<string, object> context,
-        ISmartDictionary state,
-        ISmartDictionary properties)
+            string filename, object scriptParams,
+            Dictionary<string, object> context,
+            ISmartDictionary state,
+            ISmartDictionary properties)
         {
             this.deviceState = state;
             this.deviceProperties = properties;
```
Comment (hathind-ms) [inline Services/Simulation/JavascriptInterpreter.cs]: please fix indentation

Comment (thirumathi) [reply Services/Simulation/JavascriptInterpreter.cs]: This is resolved already

File: Services/data/devicemodels/cab-01-protobuf.json (hunk 16 above)
Comment (thirumathi) [reply Services/data/devicemodels/cab-01-protobuf.json]: This is resolved. Please validate and update.

File: Services/data/devicemodels/cab-01-protobuf.json (hunk 31 above)
Comment (thirumathi) [reply Services/data/devicemodels/cab-01-protobuf.json]: Included missing class

**Comment by hathind-ms** [comment]
Protobuf support merged as part of this PR: https://github.com/Azure/device-simulation-dotnet/pull/235
